### PR TITLE
Convert with-reviews to TS

### LINKED ISSUE #42360: Convert `with-reviews.js` to `with-reviews.tsx` and replace `propTypes` with TypeScript definitions
## Problem

[`assets/js/base/hocs/with-reviews.js`](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/hocs/with-reviews.js) is a JS file and uses `propTypes` definitions.

## Expected outcome

1. Convert the file from `with-reviews.js` to `with-reviews.tsx`.
2. Replace `propTypes` with TypeScript definitions.

### PR DESCRIPTION
**Pull request by octaedro**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR transforms the HOC `with-reviews` from `js` to `tsx`.

Closes #42360.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new product and add a few reviews to it.
2. Go to Pages > Add New Page.
3. Add the blocks related to reviews: `Reviews by Category`, `Reviews by Product` and `All Reviews`.
4. They should work well.
5. Preview the page, the reviews should look well.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/base/hocs/with-reviews.tsx [hunk 1]
```diff
@@ -2,68 +2,77 @@
...
+	class WrappedComponent extends Component<
+		WithReviewsProps,
+		WithReviewsState
+	> {
 		isPreview = !! this.props.attributes.previewReviews;
 
-		delayedAppendReviews = this.props.delayFunction( this.appendReviews );
+		delayedAppendReviews = (
+			this.props.delayFunction ?? ( ( f: () => void ) => f )
+		)( this.appendReviews.bind( this ) );
```
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/base/hocs/with-reviews.tsx]: In previous code, there was no explicit binding of `this.appendReviews`. Out of curiosity, Is it necessary to make this change during JS to TSX refactor? 🤔

File: plugins/woocommerce-blocks/assets/js/base/hocs/with-reviews.tsx [hunk 2]
```diff
@@ -74,27 +83,35 @@ const withReviews = ( OriginalComponent ) => {
...
 		}
 
 		componentWillUnmount() {
 			this.isMounted = false;
-
-			if ( this.delayedAppendReviews.cancel ) {
+			if (
+				'cancel' in this.delayedAppendReviews &&
+				typeof this.delayedAppendReviews.cancel === 'function'
+			) {
```
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/base/hocs/with-reviews.tsx:106]: Thanks for adding additional checks for `cancel` method i.e. making sure it is a function 🙌🏻

**Review by imanish003 — APPROVED** [review]
Hey @octaedro,

I tested the PR and it seems to be working as expected. I left one minor question but I don't think it's a blocker so pre-approving the PR. Thanks for converting this file to TS 🚀

File: plugins/woocommerce-blocks/assets/js/base/hocs/with-reviews.tsx (hunk 1 above)
Comment (octaedro) [reply plugins/woocommerce-blocks/assets/js/base/hocs/with-reviews.tsx]: Good call! I removed the `bind` in the commit [1d515fa](https://github.com/woocommerce/woocommerce/pull/50890/commits/1d515fa227e0182d3eeac28480590681ca586b75)

**Pull request merged**